File: src/exo-engine/src/main/java/exo/engine/mapper/PodcastMapper.java
package exo.engine.mapper;

import exo.engine.domain.IndexField;
import exo.engine.domain.dto.ImmutablePodcast;
import exo.engine.domain.dto.ModifiablePodcast;
import exo.engine.domain.dto.Podcast;
import exo.engine.domain.entity.PodcastEntity;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingTarget;
import org.mapstruct.NullValueCheckStrategy;
import org.mapstruct.factory.Mappers;

import java.util.Objects;
import java.util.Optional;
import java.util.stream.Stream;

/**
 * @author Maximilian Irro
 */
@Mapper(
    uses = {UrlMapper.class, DateMapper.class},
    nullValueCheckStrategy = NullValueCheckStrategy.ALWAYS
)
public interface PodcastMapper {

    PodcastMapper INSTANCE = Mappers.getMapper( PodcastMapper.class );

    @Mapping(target = "episodes", ignore = true)
    @Mapping(target = "feeds", ignore = true)
    PodcastEntity toEntity(Podcast dto);

    ModifiablePodcast toModifiable(PodcastEntity podcast);

    default ImmutablePodcast toImmutable(PodcastEntity podcast) {
        return Optional
            .ofNullable(podcast)
            .map(this::toModifiable)
            .map(ModifiablePodcast::toImmutable)
            .orElse(null);
    }

    default ModifiablePodcast toModifiable(Podcast podcast) {
        return Optional
            .ofNullable(podcast)
            .map(p -> {
                if (p instanceof ModifiablePodcast) {
                    return (ModifiablePodcast) p;
                } else {
                    return new ModifiablePodcast().from(p);
                }})
            .orElse(null);
    }

    default ImmutablePodcast toImmutable(Podcast podcast) {
        return Optional
            .ofNullable(podcast)
            .map(p -> {
                if (p instanceof ImmutablePodcast) {
                    return (ImmutablePodcast) p;
                } else {
                    return ((ModifiablePodcast) p).toImmutable();
                }})
            .orElse(null);
    }

    ModifiablePodcast update(Podcast src, @MappingTarget ModifiablePodcast target);

    default ModifiablePodcast update(Podcast src, @MappingTarget Podcast target) {
        return Optional
            .ofNullable(target)
            .map(t -> {
                if (t instanceof ModifiablePodcast) {
                    return (ModifiablePodcast) t;
                } else {
                    return new ModifiablePodcast().from(t);
                }})
            .map(t -> update(src, t))
            .orElse(null);
    }

    default ImmutablePodcast updateImmutable(Podcast src, @MappingTarget Podcast target) {
        return Optional
            .ofNullable(target)
            .map(t -> {
                if (t instanceof ModifiablePodcast) {
                    return (ModifiablePodcast) t;
                } else {
                    return new ModifiablePodcast().from(t);
                }})
            .map(t -> update(src, t))
            .map(ModifiablePodcast::toImmutable)
            .orElse(null);
    }

    default ImmutablePodcast toImmutable(org.apache.lucene.document.Document doc) {
        return Optional
            .ofNullable(doc)
            .map(d -> ImmutablePodcast.builder()
                .setExo(d.get(IndexField.EXO))
                .setTitle(d.get(IndexField.TITLE))
                .setLink(d.get(IndexField.LINK))
                .setPubDate(Optional
                    .ofNullable(d.get(IndexField.PUB_DATE))
                    .map(DateMapper.INSTANCE::asLocalDateTime)
                    .orElse(null))
                .setDescription(Stream
                    .of(d.get(IndexField.ITUNES_SUMMARY), d.get(IndexField.DESCRIPTION))
                    .filter(Objects::nonNull)
                    .findFirst()
                    .orElse(null))
                .setImage(d.get(IndexField.ITUNES_IMAGE))
                .create())
            .orElse(null);
    }

}

File: hemin-engine/src/main/scala/hemin/engine/util/mapper/PodcastMapper.scala
package hemin.engine.util.mapper

import hemin.engine.model.{IndexDoc, IndexField, Podcast, PodcastItunes}
import hemin.engine.util.mapper.MapperErrors._
import org.apache.solr.common.SolrDocument

import scala.util.{Success, Try}

object PodcastMapper {

  def toPodcast(doc: IndexDoc): Try[Podcast] = Option(doc)
    .map{ d =>
      Podcast(
        id          = d.id,
        title       = d.title,
        link        = d.link,
        description = d.description,
        pubDate     = d.pubDate,
        image       = d.image,
        itunes = PodcastItunes(
          author  = d.itunesAuthor,
          summary = d.itunesSummary
        )
      )
    }
    .map(Success(_))
    .getOrElse(mapperFailureIndexToPodcast(doc))

  def toPodcast(doc: org.apache.lucene.document.Document): Try[Podcast] = Option(doc)
    .map { d =>
      Podcast(
        id          = LuceneMapper.get(d, IndexField.Id.entryName),
        title       = LuceneMapper.get(d, IndexField.Title.entryName),
        link        = LuceneMapper.get(d, IndexField.Link.entryName),
        pubDate     = DateMapper.asMilliseconds(d.get(IndexField.PubDate.entryName)),
        description = LuceneMapper.get(d, IndexField.Description.entryName),
        image       = LuceneMapper.get(d, IndexField.ItunesImage.entryName),
      )
    }
    .map(Success(_))
    .getOrElse(mapperFailureLuceneToPodcast(doc))


  def toPodcast(doc: SolrDocument): Try[Podcast] = Option(doc)
    .map { d =>
      Podcast(
        id          = SolrMapper.firstStringMatch(d, IndexField.Id.entryName),
        title       = SolrMapper.firstStringMatch(d, IndexField.Title.entryName),
        link        = SolrMapper.firstStringMatch(d, IndexField.Link.entryName),
        pubDate     = SolrMapper.firstLongMatch(d, IndexField.PubDate.entryName),
        description = SolrMapper.firstStringMatch(d, IndexField.Description.entryName),
        image       = SolrMapper.firstStringMatch(d, IndexField.ItunesImage.entryName),
      )
    }
    .map(Success(_))
    .getOrElse(mapperFailureSolrToPodcast(doc))

  def toTeaser(podcast: Podcast): Option[Podcast] = Option(podcast)
    .map { p =>
      Podcast(
        id = p.id,
        title = p.title,
        link = p.link,
        description = p.description,
        pubDate = p.pubDate,
        image = p.image,
        itunes = PodcastItunes(
          summary = p.itunes.summary,
        )
      )
    }

}